libcontainer/cgroups/fs2/io.go

```diff
@@ -16,32 +18,63 @@ import (
 
 func isIoSet(r *configs.Resources) bool {
 	return r.BlkioWeight != 0 ||
+		len(r.BlkioWeightDevice) > 0 ||
 		len(r.BlkioThrottleReadBpsDevice) > 0 ||
 		len(r.BlkioThrottleWriteBpsDevice) > 0 ||
 		len(r.BlkioThrottleReadIOPSDevice) > 0 ||
 		len(r.BlkioThrottleWriteIOPSDevice) > 0
 }
 
+// bfqDeviceWeightSupported checks for per-device BFQ weight support (added
+// in kernel v5.4, commit 795fe54c2a8) by reading from "io.bfq.weight".
+func bfqDeviceWeightSupported(bfq *os.File) bool {
+	if bfq == nil {
+		return false
+	}
+	_, _ = bfq.Seek(0, 0)
+	exp := []byte("default ")
```

Comment:
Is this really guaranteed to be in the first line?

Reply:
Yes; see `bfq_io_show_weight` in `block/bfq-cgroup.c`, e.g. 

https://github.com/torvalds/linux/blob/009c9aa5be652675a06d5211e1640e02bbb1c33d/block/bfq-cgroup.c#L977-L986

Reply:
And before https://github.com/torvalds/linux/commit/795fe54c2a828099e461e8c36b04210b2df462ed it was a number in there.

Reply:
Future version of the kernel may have another line before "default 100".
So I guess we have to scan all the lines?

Reply:
*In theory* that would be an ABI breakage, but yeah it seems a bit risky to depend on the exact order of lines in a kernel file when the alternative is pretty inexpensive. (And in the past cgroup ABIs haven't exactly been completely free of ABI breakages.)

Reply:
I disagree.

So far we have to distinguish between two versions of the `bfq.io.weight` API. Let's call them version 1 and version 2. Version 1 only allows to get/set the default weight. Version 2 also allows to get/set per-device weights.

Now, if the **first** line of `bfq.io.weight` does read back as `default XXX`, we do have version 2. Otherwise it's something else. Either version 1 which does not support setting per-device weights, or perhaps a version 3 we don't know anything about.

But let's assume that write format for version 3 is backward-compatible with version 2 (same as version 2 write is backward-compatible with version 1, although the read is not). In this case we need to check not for version 2, but for anything other than version 1.

Such a check would be to read the entire file and see that it's not a version 1 file, i.e. not a single number followed by a newline.

I will update the commit.

Reply:
Updated. The new check specifically checks for v1 file, not v2 file, and is thus more future-proof (assuming that in v3 file format will expand and not go back to same as it was in v1).